File: selector/src/main/java/com/luck/picture/lib/language/PictureLanguageUtils.java
package com.luck.picture.lib.language;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.util.DisplayMetrics;

import androidx.annotation.NonNull;

import com.luck.picture.lib.utils.SpUtils;
import java.util.Locale;

/**
 * @author：luck
 * @data：2018/3/28 下午1:00
 * @描述: PictureLanguageUtils
 */
public class PictureLanguageUtils {

    private static final String KEY_LOCALE = "KEY_LOCALE";
    private static final String VALUE_FOLLOW_SYSTEM = "VALUE_FOLLOW_SYSTEM";

    private PictureLanguageUtils() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * init app the language
     *
     * @param context
     * @param languageId
     * @param defaultLanguageId
     */
    public static void setAppLanguage(Context context, int languageId, int defaultLanguageId) {
        if (languageId >= 0) {
            applyLanguage(context, LocaleTransform.getLanguage(languageId));
        } else {
            if (defaultLanguageId >= 0) {
                applyLanguage(context, LocaleTransform.getLanguage(defaultLanguageId));
            } else {
                setDefaultLanguage(context);
            }
        }
    }

    /**
     * Apply the language.
     *
     * @param locale The language of locale.
     */
    private static void applyLanguage(@NonNull Context context, @NonNull final Locale locale) {
        applyLanguage(context, locale, false);
    }


    private static void applyLanguage(@NonNull Context context, @NonNull final Locale locale,
                                      final boolean isFollowSystem) {
        if (isFollowSystem) {
            SpUtils.putString(context, KEY_LOCALE, VALUE_FOLLOW_SYSTEM);
        } else {
            String localLanguage = locale.getLanguage();
            String localCountry = locale.getCountry();
            SpUtils.putString(context, KEY_LOCALE, localLanguage + "$" + localCountry);
        }
        updateLanguage(context, locale);
    }



    private static void updateLanguage(@NonNull Context context, Locale locale) {
        Resources resources = context.getResources();
        Configuration config = resources.getConfiguration();
        Locale contextLocale = config.locale;
        if (equals(contextLocale.getLanguage(), locale.getLanguage())
                && equals(contextLocale.getCountry(), locale.getCountry())) {
            return;
        }
        DisplayMetrics dm = resources.getDisplayMetrics();
        config.setLocale(locale);
        context.createConfigurationContext(config);
        resources.updateConfiguration(config, dm);
    }

    /**
     * set default language
     *
     * @param context
     */
    private static void setDefaultLanguage(Context context) {
        Resources resources = context.getResources();
        Configuration config = resources.getConfiguration();
        DisplayMetrics dm = resources.getDisplayMetrics();
        config.setLocale(Locale.getDefault());
        context.createConfigurationContext(config);
        resources.updateConfiguration(config, dm);
    }

    private static boolean equals(final CharSequence s1, final CharSequence s2) {
        if (s1 == s2) return true;
        int length;
        if (s1 != null && s2 != null && (length = s1.length()) == s2.length()) {
            if (s1 instanceof String && s2 instanceof String) {
                return s1.equals(s2);
            } else {
                for (int i = 0; i < length; i++) {
                    if (s1.charAt(i) != s2.charAt(i)) return false;
                }
                return true;
            }
        }
        return false;
    }
}

File: selector/src/main/java/com/luck/picture/lib/language/PictureLanguageUtils.kt
package com.luck.picture.lib.language

import android.content.Context
import com.luck.picture.lib.language.LocaleTransform.getLanguage
import java.util.*

/**
 * @author：luck
 * @data：2018/3/28 下午1:00
 * @描述: PictureLanguageUtils
 */
object PictureLanguageUtils {

    /**
     * init app the language
     *
     * @param context
     * @param language
     * @param defaultLanguage
     */
    fun setAppLanguage(context: Context, language: Language, defaultLanguage: Language) {
        if (language != Language.SYSTEM_LANGUAGE || language != Language.UNKNOWN_LANGUAGE) {
            applyLanguage(context, getLanguage(language))
        } else {
            if (defaultLanguage != Language.SYSTEM_LANGUAGE || defaultLanguage != Language.UNKNOWN_LANGUAGE) {
                applyLanguage(context, getLanguage(defaultLanguage))
            } else {
                setDefaultLanguage(context)
            }
        }
    }

    /**
     * Apply the language.
     *
     * @param locale The language of locale.
     */
    private fun applyLanguage(context: Context, locale: Locale) {
        val resources = context.resources
        val config = resources.configuration
        val contextLocale = config.locale
        if (equals(contextLocale.language, locale.language) && equals(
                contextLocale.country,
                locale.country
            )
        ) {
            return
        }
        val dm = resources.displayMetrics
        config.setLocale(locale)
        context.createConfigurationContext(config)
        resources.updateConfiguration(config, dm)
    }

    /**
     * set default language
     *
     * @param context
     */
    private fun setDefaultLanguage(context: Context) {
        val resources = context.resources
        val config = resources.configuration
        val dm = resources.displayMetrics
        config.setLocale(Locale.getDefault())
        context.createConfigurationContext(config)
        resources.updateConfiguration(config, dm)
    }

    private fun equals(s1: CharSequence?, s2: CharSequence?): Boolean {
        if (s1 === s2) return true
        var length = 0
        return if (s1 != null && s2 != null && s1.length.also { length = it } == s2.length) {
            if (s1 is String && s2 is String) {
                s1 == s2
            } else {
                for (i in 0 until length) {
                    if (s1[i] != s2[i]) return false
                }
                true
            }
        } else false
    }
}